### PR TITLE
Fix RTLSDR serial argument parsing

### PR DESCRIPTION
**Pull request by f00b4r0**
Hi,

This PR addresses the following issue:

- First, the current way `--rtlsdr` parses its argument makes it impossible to disambiguate explicit serial numbers with low "value" (e.g. `00000001`, which is interpreted by the current code as `--rtlsdr 1` selecting device ID 1 instead of device with SN `00000001`). The proposed commit will always parse any 8-character-long argument as a serial, preserving existing behaviour for all other cases (non-8c-long arguments). There are other ways to rework this part of the code but I wanted to keep things as simple and as backward-compatible as possible.
- ~~Second, an additional commit is proposed to add signal stats reporting on good decoded messages via statsd (signal power, noise floor and ppm error). This is useful to keep track of the performance of the dongle and to check improvements due to antenna adjustments.~~

I hope this helps

### CODE REVIEW

**Comment by szpajder** [comment]
Thank you, but the statsd part doesn't seem right. `v->frame_pwr` is a very small value (less than 1), so `log10f` of it is always negative. How is it supposed to cast correctly to `size_t` which is an unsigned type?

What's more, sending these stats inside the `while((ret = bitstream_copy_next_frame...` loop is redundant, because signal levels and PPM error are computed once per VDL2 burst. If a burst contains more than one VDL2 frame, the program will send the same set values more than once. IMO sending them once per burst (ie. outside of the loop) is therefore sufficient.

Also please resubmit the PR against the unstable branch, since this is where the development is done. master is for doing releases only. Thanks in advance.

**Comment by f00b4r0** [comment]
> Thank you, but the statsd part doesn't seem right. `v->frame_pwr` is a very small value (less than 1), so `log10f` of it is always negative. How is it supposed to cast correctly to `size_t` which is an unsigned type?

It doesn't "per se", but since the statsd-c-client implementation [uses `%zd`](https://github.com/romanbsd/statsd-c-client/blob/ef6f82f1a2d77c5ef73272ae87a0670b1b81f005/statsd-client.c#L230) printf format, it Just Work™. The values are correctly reported as negative integers. So while it's not _exactly_ correct, I'd leave it as is because it does what's expected and works with the inadequacies of the statsd-c-client API (which requires a `size_t` argument for gauges).

That statsd implementation is woefully inefficient by the way (sending one UDP packet for every counter increase, etc): would you be interested in a better approach for the entire stats thing? (and more generally are you interested in performance improvements patches?)

> What's more, sending these stats inside the `while((ret = bitstream_copy_next_frame...` loop is redundant, because signal levels and PPM error are computed once per VDL2 burst. If a burst contains more than one VDL2 frame, the program will send the same set values more than once. IMO sending them once per burst (ie. outside of the loop) is therefore sufficient.

Ah ok, good to know. That's an easy fix.

> Also please resubmit the PR against the unstable branch, since this is where the development is done. master is for doing releases only. Thanks in advance.

OK. I may have missed that bit of info then, it's unusual to have `master` not being the main development branch :)

Will resubmit.

**Comment by f00b4r0** [comment]
ok, changes applied and rebased onto `unstable`. So as not to lose the PR history I changed the PR base branch instead of closing/reopening. HTH

**Comment by szpajder** [comment]
> > Thank you, but the statsd part doesn't seem right. `v->frame_pwr` is a very small value (less than 1), so `log10f` of it is always negative. How is it supposed to cast correctly to `size_t` which is an unsigned type?
> 
> It doesn't "per se", but since the statsd-c-client implementation [uses `%zd`](https://github.com/romanbsd/statsd-c-client/blob/ef6f82f1a2d77c5ef73272ae87a0670b1b81f005/statsd-client.c#L230) printf format, it Just Work™. The values are correctly reported as negative integers. So while it's not _exactly_ correct, I'd leave it as is because it does what's expected and works with the inadequacies of the statsd-c-client API (which requires a `size_t` argument for gauges).

But it doesn't work. At least on a 32-bit Arch Linux on a Raspberry Pi it's sending only zeros.

```
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on wlan0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
E..MI.@.@.l....^.....n...9'%dumpvdl2.136725000.decoder.msg.good_pwr_mdbfs:0|g
E..LJ.@.@.l....^.....n...8.>dumpvdl2.136725000.decoder.msg.good_nf_mdbfs:0|g
E..MJ.@.@.l....^.....n...9'%dumpvdl2.136725000.decoder.msg.good_pwr_mdbfs:0|g
E..LJ.@.@.l....^.....n...8.>dumpvdl2.136725000.decoder.msg.good_nf_mdbfs:0|g
E..MJ+@.@.l....^.....n...9' dumpvdl2.136775000.decoder.msg.good_pwr_mdbfs:0|g
E..LJ,@.@.l....^.....n...8.9dumpvdl2.136775000.decoder.msg.good_nf_mdbfs:0|g
E..MJA@.@.l....^.....n...9% dumpvdl2.136975000.decoder.msg.good_pwr_mdbfs:0|g
E..LJB@.@.l....^.....n...8.9dumpvdl2.136975000.decoder.msg.good_nf_mdbfs:0|g
E..MJU@.@.l....^.....n...9'%dumpvdl2.136725000.decoder.msg.good_pwr_mdbfs:0|g
E..LJW@.@.l....^.....n...8.>dumpvdl2.136725000.decoder.msg.good_nf_mdbfs:0|g
E..MJ.@.@.l]...^.....n...9'%dumpvdl2.136725000.decoder.msg.good_pwr_mdbfs:0|g
E..LJ.@.@.l]...^.....n...8.>dumpvdl2.136725000.decoder.msg.good_nf_mdbfs:0|g
E..MJ.@.@.lN...^.....n...9'%dumpvdl2.136725000.decoder.msg.good_pwr_mdbfs:0|g
E..LJ.@.@.lN...^.....n...8.>dumpvdl2.136725000.decoder.msg.good_nf_mdbfs:0|g
E..MJ.@.@.l....^.....n...9' dumpvdl2.136775000.decoder.msg.good_pwr_mdbfs:0|g
E..LJ.@.@.l....^.....n...8.9dumpvdl2.136775000.decoder.msg.good_nf_mdbfs:0|g
E..MJ.@.@.k....^.....n...9'%dumpvdl2.136725000.decoder.msg.good_pwr_mdbfs:0|g
E..LJ.@.@.k....^.....n...8.>dumpvdl2.136725000.decoder.msg.good_nf_mdbfs:0|g
```

OTOH PPM values are sent correctly.

> That statsd implementation is woefully inefficient by the way (sending one UDP packet for every counter increase, etc): would you be interested in a better approach for the entire stats thing?

statsd is meant to be simple and not overengineered. It serves the purpose and its overhead is negligible.

**Comment by f00b4r0** [comment]
> > > Thank you, but the statsd part doesn't seem right. `v->frame_pwr` is a very small value (less than 1), so `log10f` of it is always negative. How is it supposed to cast correctly to `size_t` which is an unsigned type?
> > 
> > 
> > It doesn't "per se", but since the statsd-c-client implementation [uses `%zd`](https://github.com/romanbsd/statsd-c-client/blob/ef6f82f1a2d77c5ef73272ae87a0670b1b81f005/statsd-client.c#L230) printf format, it Just Work™. The values are correctly reported as negative integers. So while it's not _exactly_ correct, I'd leave it as is because it does what's expected and works with the inadequacies of the statsd-c-client API (which requires a `size_t` argument for gauges).
> 
> But it doesn't work. At least on a 32-bit Arch Linux on a Raspberry Pi it's sending only zeros.


Here it works (x86-64 Debian):

```
..<...G*bdumpvdl2.136975000.decoder.msg.good_pwr_mdbfs:-10818|g
..<...F*adumpvdl2.136975000.decoder.msg.good_nf_mdbfs:-28872|g
..<...G*bdumpvdl2.136975000.decoder.msg.good_mppm_error:-2113|g
```

I'll try to reproduce on 32-bit armhf.

> OTOH PPM values are sent correctly.

hold on, the *negative* ppm values are sent correctly?

> > That statsd implementation is woefully inefficient by the way (sending one UDP packet for every counter increase, etc): would you be interested in a better approach for the entire stats thing?
> 
> statsd is meant to be simple and not overengineered. It serves the purpose and its overhead is negligible.

I guess it depends what you call "negligible" and how many instances of dumpvdl2 you're feeding to your statsd receiver :)
But ok, if you're not interested I won't waste my time.

**Comment by f00b4r0** [comment]
>> statsd is meant to be simple and not overengineered. It serves the purpose and its overhead is negligible.

> I guess it depends what you call "negligible" and how many instances of dumpvdl2 you're feeding to your statsd receiver :)

BTW I'm not saying that statsd is bad (although it isn't great ;): I use it in some of my code, I just don't use statsd-c-client which is a terrible implementation. Just to be clear I was proposing to rework that part, not getting rid of statsd as an output protocol.

**Comment by f00b4r0** [comment]
OK I dropped the stats patch. The statsd-c-client implementation is broken on many accounts and will not handle negative values (gauges or otherwise) due to its use of `size_t`. I suspect [`statsd_dec()`](https://github.com/romanbsd/statsd-c-client/blob/ef6f82f1a2d77c5ef73272ae87a0670b1b81f005/statsd-client.c#L242) only happens to work because [`%zd`](https://github.com/romanbsd/statsd-c-client/blob/ef6f82f1a2d77c5ef73272ae87a0670b1b81f005/statsd-client.c#L230) accepts a `ssize_t` value, and the POSIX standards defines `ssize_t ` as:

```
ssize_t
              Used for a count of bytes or an error indication.  It  is  a  signed  integer  type
              capable of storing values al least in the range [-1, SSIZE_MAX].
```

So `-1` in the implicit cast of the `size_t` argument works.

The reason the patch works for me on x86-64 Debian and not for you on armhf Arch Linux is probably due to compiler/glibc/arch implementation differences; looking at the output of `objdump` would probably clarify this but I'm too lazy to waste more time on this since this is unfixable within dumpvdl2.

Furthermore, this reminded me that statsd broken design (or lack thereof) also does not support standalone negative gauges, they need to be first set to zero and then the negative value must be sent (at least in [some implementations](https://github.com/statsd/statsd/blob/master/docs/metric_types.md#gauges)). I modified locally `statsd_gauge_set()` to do that and now I get more sensible values for SNR and PPM compared to the odd ones I was observing before.

tl;dr: the correct fix is to NOT use statsd-c-client, but since you seem uninterested in such a patch I guess it's not worth the effort. So this PR is now reduced to the rtlsdr serial fix, I hope this'll be accepted.

Cheers

**Comment by szpajder** [comment]
Patch merged, thanks.

I just remembered that I've implemented noise floor reporting over statsd in dumpfdl (see [here](https://github.com/szpajder/dumphfdl/blob/61cec767f09119f533f9dfe770cc8f802479182f/src/hfdl.c#L1093)). It sends the absolute noise floor value expressed in tenths of dB. In Grafana I plot these values scaled by -10.

**Comment by f00b4r0** [comment]
> Patch merged, thanks.

Thanks

> I just remembered that I've implemented noise floor reporting over statsd in dumpfdl (see [here](https://github.com/szpajder/dumphfdl/blob/61cec767f09119f533f9dfe770cc8f802479182f/src/hfdl.c#L1093)). It sends the absolute noise floor value expressed in tenths of dB. In Grafana I plot these values scaled by -0.1

Yeah that's really a bandaid though :) (And I see you have the same comment about negative gauges.)
Also the values I'm most interested in are SNR (so I need signal strength) and PPM (to monitor a potentially drifting dongle).

If you would consider moving away from statsd-c-client I'd be tempted to switch to a bespoke implementation (I did this for [my own code](https://github.com/f00b4r0/rwchcd/blob/master/log/log_statsd.c), so I have a good understanding of what's needed - after all we're only dealing with strings sent over UDP, it's not rocket science), but only if this is something you would consider merging. I'd rather not waste my time otherwise :)

**Comment by szpajder** [comment]
Bandaid or not, it makes the job done in a single line of code. KISS rule at its best. The less code, the better.

**Comment by f00b4r0** [comment]
> Bandaid or not, it makes the job done in a single line of code. KISS rule at its best. The less code, the better.

I agree. It just won't work for PPM for instance.
